### PR TITLE
12 chore add tests for request utils

### PR DESCRIPTION
**Pull request by alexander-ma**
# Summary

### Related Issues

- Completes #12 

### Contributors

- @alexander-ma : implementer

### Affected Modules

- `pkg/util/requests`

### Description
Added test coverage as a learning exercise for `requests.go` and implementing the testify library for assertions.

# Testing (Unit/Integration/Performance)

Added tests to cover basic usage of `request.go` (GET, POST, Custom)

# Points of Attention

This test class covers a very basic amount of use cases and does not cover edge cases. Additional tests may be added to improve code coverage.

### CODE REVIEW

File: pkg/utils/requests/requests_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	testServer := httptest.NewServer(getTestHandler)
+	params := map[string]string{"param1": "foobar"}
+	_, err := GetRequest(testServer.URL, &params, nil)
+	require.NoError(t, err)
+}
+
+func TestRequests_GetRequest_NilParams(t *testing.T) {
+	getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, r.Method, http.MethodGet, "Request is not for GET Method")
+		r.ParseForm()
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:33]: lint - check error return

File: pkg/utils/requests/requests_test.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+	testServer := httptest.NewServer(getTestHandler)
+	_, err := GetRequest(testServer.URL, nil, nil)
+	require.NoError(t, err)
+}
+
+func TestRequests_PostRequest(t *testing.T) {
+	getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, r.Method, http.MethodPost, "Request is not for POST Method")
+		r.ParseForm()
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:45]: lint - check error return

File: pkg/utils/requests/requests_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	testServer := httptest.NewServer(getTestHandler)
+	params := map[string]string{"param1": "foobar"}
+	_, err := PostRequest(testServer.URL, "application/json", &params, nil, nil)
+	require.NoError(t, err)
+}
+
+func TestRequest_CustomRequest(t *testing.T) {
+	getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, r.Method, "custom", "Request is not for custom method")
+		r.ParseForm()
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:59]: lint - check error return

File: pkg/utils/requests/requests_test.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// Full test coverage for GetRequest, PostRequest
+// CustomRequest can be covered using two different http methods (e.g. PUT, DELETE)
+// Test can be housed in pkg/utils/requests/requests_test.go
+// Separate test functions for each wrapper tested, use sub tests if needed
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:16]: These sound like preliminary notes rather than describing the actual tests

File: pkg/utils/requests/requests_test.go [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+	})
+
+	testServer := httptest.NewServer(getTestHandler)
+	params := map[string]string{"param1": "foobar"}
+	_, err := GetRequest(testServer.URL, &params, nil)
+	require.NoError(t, err)
+}
+
+func TestRequests_GetRequest_NilParams(t *testing.T) {
+	getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:31]: I'd use two subtests combining the `GET` tests, something like:

```go
func TestRequests_GETRequest(t *testing.T) {
    t.Run("basic", func() {
        ...
    }

    t.Run("nil params", func() {
        ...
    }
}
```

File: pkg/utils/requests/requests_test.go (hunk 4 above)
Comment (alexander-ma) [reply pkg/utils/requests/requests_test.go]: Removed these comments as they were for my notes.

File: pkg/utils/requests/requests_test.go (hunk 5 above)
Comment (alexander-ma) [reply pkg/utils/requests/requests_test.go]: This is a good suggestion, thanks.

File: pkg/utils/requests/requests_test.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestRequests_GetRequest(t *testing.T) {
+	t.Run("basic", func(t *testing.T) {
+		getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			assert.Equal(t, r.Method, http.MethodGet, "Request is not for GET Method")
+			assert.Equal(t, r.URL.Query().Get("param1"), "foobar", "Param values do not match")
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:17]: _nit_  - u can extra tryhard this shit by setting consts for `param1` & `foobar` at the start - this makes it easier to change all the references at once up 2 u tho

File: pkg/utils/requests/requests_test.go [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+
+		testServer := httptest.NewServer(getTestHandler)
+		params := map[string]string{}
+		_, err := PostRequest(testServer.URL, "application/json", &params, nil, nil)
+		if assert.Error(t, err) {
+			assert.Equal(t, "no parameters provided", err.Error())
+		}
+	})
+	t.Run("default case", func(t *testing.T) {
+		getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:71]: no testing at all?

File: pkg/utils/requests/requests_test.go [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+			assert.Equal(t, http.MethodPost, r.Method, "Request is not for POST Method")
+			err := r.ParseForm()
+			require.NoError(t, err)
+			assert.Empty(t, r.Form)
+		})
+
+		testServer := httptest.NewServer(getTestHandler)
+		params := map[string]string{"param1": "foobar"}
+		value := []byte("value")
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:104]: no assertions on value or uploadData? this is pretty important

File: pkg/utils/requests/requests_test.go [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+	})
+	t.Run("application/x-www-form-urlencoded", func(t *testing.T) {
+		getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			assert.Equal(t, http.MethodPost, r.Method, "Request is not for POST Method")
+			err := r.ParseForm()
+			require.NoError(t, err)
+		})
+
+		testServer := httptest.NewServer(getTestHandler)
+		params := map[string]string{}
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:89]: whats the purpose of passing `string{}`? also no assertions?

**Review by dvo-dev — CHANGES_REQUESTED** [review]

File: pkg/utils/requests/requests_test.go [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+	t.Run("application/json", func(t *testing.T) {
+		getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			assert.Equal(t, http.MethodPost, r.Method, "Request is not for POST Method")
+			err := r.ParseForm()
+			require.NoError(t, err)
+			assert.Empty(t, r.Form)
+		})
+
+		testServer := httptest.NewServer(getTestHandler)
+		params := map[string]string{"param1": "foobar"}
```
[the remark this answers is not in the document]
Comment (alexander-ma) [reply pkg/utils/requests/requests_test.go:55]: Added assertion

File: pkg/utils/requests/requests_test.go (hunk 8 above)
Comment (alexander-ma) [reply pkg/utils/requests/requests_test.go:118]: Updated unit test to assert

File: pkg/utils/requests/requests_test.go (hunk 7 above)
Comment (alexander-ma) [reply pkg/utils/requests/requests_test.go:79]: Updated.

File: pkg/utils/requests/requests_test.go (hunk 9 above)
Comment (alexander-ma) [reply pkg/utils/requests/requests_test.go]: Removed unused params map and added assertions.

File: pkg/utils/requests/requests_test.go (hunk 6 above)
Comment (alexander-ma) [reply pkg/utils/requests/requests_test.go:19]: adding tech debt

File: pkg/utils/requests/requests_test.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+		getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			assert.Equal(t, http.MethodPost, r.Method, "Request is not for POST Method")
+			err := r.ParseMultipartForm(2 << 20)
+			require.NoError(t, err)
+			assert.Equal(t, "foobar", r.MultipartForm.Value["param1"][0])
+		})
+
+		testServer := httptest.NewServer(getTestHandler)
+		params := map[string]string{"param1": "foobar"}
+		value := []byte("value")
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:117]: are you actually checking for this value in the test handler?

**Review by dvo-dev — CHANGES_REQUESTED** [review]

File: pkg/utils/requests/requests_test.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+			assert.Equal(t, "unsupported content type", err.Error())
+		}
+	})
+	t.Run("application/x-www-form-urlencoded", func(t *testing.T) {
+		getTestHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			assert.Equal(t, http.MethodPost, r.Method, "Request is not for POST Method")
+			headers := r.Header
+			assert.NotNil(t, headers)
+			assert.Equal(t, "application/x-www-form-urlencoded", headers["Content-Type"][0])
+			assert.Equal(t, "13", headers["Content-Length"][0])
```
Comment (dvo-dev) [inline pkg/utils/requests/requests_test.go:95]: what is this 13 value?

Comment (alexander-ma) [reply pkg/utils/requests/requests_test.go:95]: 13 will come from `r.ContentLength`

File: pkg/utils/requests/requests_test.go (hunk 11 above)
Comment (alexander-ma) [reply pkg/utils/requests/requests_test.go:123]: Added check to look for byte array.

**Review by dvo-dev — APPROVED** [review]
LGTM

**Comment by alexander-ma** [comment]
Pog
